core/src/main/java/org/apache/iceberg/Partition.java

```diff
@@ -0,0 +1,355 @@
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    } else if (!(o instanceof Partition)) {
+      return false;
+    }
+
+    Partition that = (Partition) o;
+    return partitionData.equals(that.partitionData)
+        && specId == that.specId
+        && dataRecordCount == that.dataRecordCount
+        && dataFileCount == that.dataFileCount
+        && dataFileSizeInBytes == that.dataFileSizeInBytes
+        && posDeleteRecordCount == that.posDeleteRecordCount
+        && posDeleteFileCount == that.posDeleteFileCount
+        && eqDeleteRecordCount == that.eqDeleteRecordCount
+        && eqDeleteFileCount == that.eqDeleteFileCount
+        && totalRecordCount == that.totalRecordCount
+        && lastUpdatedAt == that.lastUpdatedAt
+        && lastUpdatedSnapshotId == that.lastUpdatedSnapshotId;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(
+        partitionData,
+        specId,
+        dataRecordCount,
+        dataFileCount,
+        dataFileSizeInBytes,
+        posDeleteRecordCount,
+        posDeleteFileCount,
+        eqDeleteRecordCount,
+        eqDeleteFileCount,
+        totalRecordCount,
+        lastUpdatedAt,
+        lastUpdatedSnapshotId);
+  }
+
+  public static org.apache.iceberg.Schema icebergSchema(Types.StructType partitionType) {
+    if (partitionType.fields().isEmpty()) {
+      throw new IllegalArgumentException("getting schema for an unpartitioned table");
+    }
+
+    return new org.apache.iceberg.Schema(
+        Types.NestedField.required(1, Column.PARTITION_DATA.name(), partitionType),
+        Types.NestedField.required(2, Column.SPEC_ID.name(), Types.IntegerType.get()),
+        Types.NestedField.required(3, Column.DATA_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.required(4, Column.DATA_FILE_COUNT.name(), Types.IntegerType.get()),
+        Types.NestedField.required(5, Column.DATA_FILE_SIZE_IN_BYTES.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            6, Column.POSITION_DELETE_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            7, Column.POSITION_DELETE_FILE_COUNT.name(), Types.IntegerType.get()),
+        Types.NestedField.optional(
+            8, Column.EQUALITY_DELETE_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            9, Column.EQUALITY_DELETE_FILE_COUNT.name(), Types.IntegerType.get()),
+        Types.NestedField.optional(10, Column.TOTAL_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.optional(11, Column.LAST_UPDATED_AT.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            12, Column.LAST_UPDATED_SNAPSHOT_ID.name(), Types.LongType.get()));
+  }
+
+  private static Schema prepareAvroSchema(Types.StructType partitionType) {
+    return AvroSchemaUtil.convert(icebergSchema(partitionType), "partitionEntry");
+  }
+
+  public static class Builder {
```

Comment:
using the builder instead of `Immutables` because. These objects will be mutable in a partition stats map. Will be adding the update functions during impl.